analysis/analysis.py

```diff
@@ -0,0 +1,297 @@
+        if pred(v):
+            return i
+    return None
+
+
+def has_more_specific_task_bracket(task_bracket, task_indices):
+    for other in task_indices:
+        if other == task_bracket:
+            continue
+        if other[0] in range(task_bracket[0], task_bracket[1]):
+            # theoretically other[1] == end should also be checked for inclusiveness
+            # however we expect to have a tree here where this should not happen
+            return True
+    return False
+
+
+def append_subtask(main_task_name, table_rows, csv_rows, subtasks):
+    ids = set(map(lambda t: t[2]["id"], subtasks))
+    tasks_length = len(subtasks)
+    for id in ids:
+        filtered_tasks = []
+        for i in range(tasks_length):
+            if "id" not in subtasks[i][2]:
+                continue
+            if id == subtasks[i][2]["id"]:
+                filtered_tasks.append(subtasks[i])
+
+        joined_content = "<br>------------------------------------<br>".join(
+            map(
+                lambda t: json.dumps(t, sort_keys=True, indent=4).replace("\n", "<br>"),
+                filtered_tasks,
+            )
+        )
+        start = filtered_tasks[0][0]
+        finish = filtered_tasks[-1][0]
+        table_rows.append(
+            {
+                "id": id,
+                "name": "Subtasks(#" + id + ")",
+                "duration": calculate_duration(start, finish),
+                "description": joined_content,
+            }
+        )
+        csv_rows.append(
+            [main_task_name, "Subtasks(#" + id + ")", calculate_duration(start, finish), ""]
+        )
+
+
+def calculate_duration(start, finish):
+    start_datetime = datetime.strptime(start, "%Y-%m-%d %H:%M:%S.%f")
+    finish_datetime = datetime.strptime(finish, "%Y-%m-%d %H:%M:%S.%f")
+    return finish_datetime - start_datetime
+
+
+def create_task_brackets(stripped_content):
+    task_indices = []
+    for i, start_content in enumerate(stripped_content):
+        start_event = start_content[1]
+        start_id = start_content[2]["id"]
+        if re.match("starting task", start_event, re.IGNORECASE):
+            for j, following_content in enumerate(stripped_content[i:], start=i):
+                following_event = following_content[1]
+                following_id = following_content[2]["id"]
+                if "task" not in following_content[2]:
+                    continue
+                ids_identical = start_id == following_id
+                event_matches_successful = re.match(
+                    "task successful", following_event, re.IGNORECASE
+                )
+                event_matches_errored = re.match("task errored", following_event, re.IGNORECASE)
+                if ids_identical and (event_matches_successful or event_matches_errored):
+                    task_indices.append([i, j])
+                    break
+    return task_indices
+
+
+def read_raw_content(input_file):
+    content = []
+    if not input_file:
```

Comment:
I had to do that in order to pipe a log from `STDIN` into the analysis, as mentioned/requested by @nlsdfnbch. Not sure if this can be improved there anyway.